### PR TITLE
fix(nef): NEF lock scanner edge cases

### PR DESCRIPTION
**Pull request by dcarley**
## Proposed Changes

Fix a handful of issues with the NEF lock scanner.

Everything in [ECO-134](https://linear.app/floxdotdev/issue/ECO-134/nef-lock-catalog-scan-validate-edge-cases-against-catalog-server) and the sub-issues except:

- [ECO-168](https://linear.app/floxdotdev/issue/ECO-168/nef-scan-imports-via-callpackage-and-with-are-not-followed)
- [ECO-166](https://linear.app/floxdotdev/issue/ECO-166/bare-dynamic-owner-catalog-reference-is-unresolvable-not-a-wildcard) server-side changes

This is more code and more LLM-ed than I would normally have bitten off in one go but:

- the original code was LLM-ed with minimal human review so I don't feel terribly attached to it
- it's worth mopping up the edge cases that we can easily find and fix now before someone runs into them
- I've gone through several review rounds and tried to anchor them all with realistic scenarios
- test coverage is entirely additive rather than modified

Warnings aren't currently user visible via `flox build` or `flox publish`, even with verbose flags, but we can consider exposing them in [ECO-133](https://linear.app/floxdotdev/issue/ECO-133/nef-lock-catalog-scan-fail-loudly-on-parse-io-errors) or later.

## Release Notes

N/A, not released as generally available

### CODE REVIEW

File: cli/nef-lock-catalog/src/bin/lock.rs
Comment (dcarley) [inline cli/nef-lock-catalog/src/bin/lock.rs:1]: Having gone 3~ rounds of local review and adjustment already I'm inclined to punt on the remaining issues that the Claude review has found but would welcome a second opinion.

File: cli/nef-lock-catalog/src/lock/lookup.rs [hunk 1]
```diff
@@ -39,6 +39,16 @@ pub enum LockError {
     #[error("{} catalog reference(s) were unresolvable", .0.len())]
     Unresolvable(Vec<UnresolvableEntry>),
 
+    /// The scan produced a root-level sentinel (`catalogs.*`): the expression
+    /// uses the whole catalog namespace in a way that cannot be analyzed
+    /// statically (e.g. passes `catalogs` to an opaque function or selects a
+    /// catalog dynamically). The wire form of such a reference (`*`) can
+    /// never resolve, so fail before making a request.
+    #[error(
+        "Cannot determine which catalogs the expression uses; it references the whole catalog namespace in a way that cannot be analyzed.\nReference packages as 'catalogs.<catalog>.<package>', or forward 'catalogs' only through 'import' arguments."
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/lock/lookup.rs]: suggestion:

```
        "Cannot determine which catalogs the expression uses; it references the whole catalog namespace in a way that cannot be analyzed.\n\
         Reference packages as 'catalogs.<catalog>.<package>', or forward 'catalogs' only through 'import' arguments."
```

File: cli/nef-lock-catalog/src/scan/mod.rs
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs:1]: fwiw with net 2.3k new lines, plus whatever it had before (even if we consider this "generated dont touch") i think its time to split this up into some reasonable pieces.

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 2]
```diff
@@ -184,26 +320,32 @@ fn line_col(content: &str, offset: usize) -> (usize, usize) {
...
+/// file's location, used only for verbose reference reporting. `root_origins`
+/// maps each root to the top-level root it stands for — the identity map at
+/// the entry file, and the composition of the forwarding chain in imported
+/// files.
 fn analyze_file_at(
     content: &str,
     roots: &HashSet<String>,
     file_dir: Option<&Path>,
-    visited: &mut HashSet<PathBuf>,
+    visited: &mut HashSet<(PathBuf, BTreeMap<String, String>)>,
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs]: a set of maps?, should this become a *map* now?

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 3]
```diff
@@ -49,6 +50,70 @@ impl Display for CatalogRef {
...
+    /// the forwarded namespaces cannot be discovered, so the scan fails
+    /// rather than silently under-locking.
+    #[error("{}", unreadable_import_message(target, file.as_deref(), *position))]
+    UnreadableImport {
+        target: PathBuf,
+        file: Option<PathBuf>,
+        /// 1-based `(line, column)` of the import application.
+        position: (usize, usize),
+    },
+}
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs:79]: if you believe the lemmings (and my intuition), outside of testsm we should practically never not know the file we're scanning outside of tests.

optionality for tests in the error schema (and transitive functions), seems a bit leaky

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 4]
```diff
@@ -184,26 +320,32 @@ fn line_col(content: &str, offset: usize) -> (usize, usize) {
...
     path: Option<&Path>,
-) -> FileInfo {
+    root_origins: &BTreeMap<String, String>,
+) -> Result<FileInfo, ScanError> {
     if let Some(path) = path {
         debug!(file = %path.display(), "reading NEF expression");
     }
 
     let parse = rnix::Root::parse(content);
     let root = parse.tree();
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs:302]: q: wasn't one of the original issues that we ignored parsing errors exactly here?

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 5]
```diff
@@ -216,21 +358,188 @@ fn analyze_file_at(
...
+    if let Some(expr) = root.expr() {
+        walker.walk(expr.syntax(), &env);
+    }
 
+    let Walker {
+        mut refs,
+        pending_imports,
+        first_root_use,
+        ..
+    } = walker;
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs]: structurally this is a bit curious, having to declare these working fields just to then extract them out again.

I'm not sure if this ever needs to be constructed partially initialized but an interface like

```
Walker::new(roots, declared_params, ctx).walk().finish() -> ( refs, imports, first_root_use ) // < what ever these mean...
```

might be clearer without carrying the whole walker structure around.

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 6]
```diff
@@ -216,21 +358,188 @@ fn analyze_file_at(
...
+    // Imports are IO: the walker only records facts, the drain here reads and
+    // recurses. Relative paths resolve against the importing file's directory.
     if let Some(dir) = file_dir {
-        follow_imports(root.syntax(), roots, &aliases, dir, visited, &mut refs);
+        for pending in pending_imports {
+            let target = dir.join(&pending.path);
+            // The fallback for an uncanonicalizable (missing) target still
+            // normalizes `.` components so errors show a clean path.
+            let target =
+                fs::canonicalize(&target).unwrap_or_else(|_| target.components().collect());
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs:367]: nit: we could use CanonicalPath as a clearer interface across here.

**Review by ysndr — COMMENTED** [review]
first round of comments, I'll take a stab at letting my lemmings help me navigate this module and finish the review tomorrow.

File: cli/nef-lock-catalog/src/scan/mod.rs (hunk 4 above)
Comment (dcarley) [reply cli/nef-lock-catalog/src/scan/mod.rs:302]: Do you mean https://linear.app/floxdotdev/issue/ECO-133/nef-lock-catalog-scan-fail-loudly-on-parse-io-errors?

This PR errors on import file reads but other missing and unparseable files, plus potentially surfacing warnings, belong in that ticket.

File: cli/nef-lock-catalog/src/scan/mod.rs (hunk 6 above)
Comment (dcarley) [reply cli/nef-lock-catalog/src/scan/mod.rs:367]: Can you expand on "across here" - do you mean within these few lines of code and map the errors to `UnreadableImport`, or further out? Can you provide a diff?

File: cli/nef-lock-catalog/src/scan/mod.rs (hunk 4 above)
Comment (ysndr) [reply cli/nef-lock-catalog/src/scan/mod.rs:302]: yeah i think thats the part of 133 i was referring to.

I got a bit confused by that as error handling was mostly addressed here and thats my context from when i last looked in here.

I agree we should handle that in the dedicated ticket.

File: cli/nef-lock-catalog/src/scan/mod.rs (hunk 6 above)
Comment (ysndr) [reply cli/nef-lock-catalog/src/scan/mod.rs:367]: I mean using `CanonicalPath` throughout, starting here, passing it to analyze_file_at, and wherever analyze file takes it.

Its a bit on the pedantic side, but if the file exisiting is an assumption of analyze_file_at we could as well enforce it via the type (which we already have for that purpose)

File: cli/nef-lock-catalog/test_data/catalog_refs/import-entry-unreadable.nix [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Pattern: the import target does not exist, so the refs it would contribute
+# through the forwarded namespace cannot be discovered and the scan fails.
+{ catalogs }:
+import ./no-such-helper.nix { inherit catalogs; }
```
Comment (ysndr) [inline cli/nef-lock-catalog/test_data/catalog_refs/import-entry-unreadable.nix:4]: does it need to *fail*?

this would also fail for

```
if false then import ./no-such-helper.nix { inherit catalogs; } else {}
```

it also *likely* fails for `import <nixpkgs>`

**Review by ysndr — APPROVED** [review]
I'm going by the new test cases for a review of the walker stuff.

those cases look pretty good.

I want to address splitting out and my comment on the graph building, but that can be done as a follow up while youre out

File: cli/nef-lock-catalog/src/scan/mod.rs [hunk 8]
```diff
@@ -106,7 +164,7 @@ pub fn scan_package_with_roots(
     base_dir: impl AsRef<Path>,
     rel_file: impl AsRef<Path>,
     roots: impl IntoIterator<Item = impl Into<String>>,
-) -> BTreeSet<CatalogRef> {
+) -> Result<BTreeSet<CatalogRef>, ScanError> {
```
Comment (ysndr) [inline cli/nef-lock-catalog/src/scan/mod.rs:167]: @dan ive been thinking about mechanism of scan package, because it kept confusing me. how its going back into files to resolve more dependencies.

it turns out that we have overloaded "dependencies" here, as "catalog dependencies" tracked by analyze_file, which yields catalog refs and "package dependencies" i.e potential local dependencies / definitions of packages.

I'm also unhappy with the opaque db hashmap, and how `collect_transitive` both expands the closure *and* extracts references.

so besides a few renamings, i'm also considering suggesting making that graph computation more explicit:

```
  graph = PackageGraph::new(base_dir, root_attributes) 
  graph.add_root(path)
  graph.expand_closure()
  let references = graph.references()
```

I've passed that critique to claude for formatting purposes and estimation below.

I know ypur not terribly attached to this but i'd like to check with you as ill likely put that forward next

---

**Naming + structure cleanup for** `scan/mod.rs`

The scanner traverses two edge kinds that both read as "dependency," the closure logic mixes graph-building with reference extraction, and `refs` stay `String` until the very last line. Proposed cleanup — all internal to `scan/mod.rs`; the public surface is only the four re-exports in `lib.rs:30` and none change.

**Renames**


| Current | Proposed | Why |
| -- | -- | -- |
| `FileInfo.deps` / `dep_names` | `dependency_args` / `dependency_arg_names` | The package function's formal arguments resolved as sibling packages — distinct from `import` edges, which keep the `import` name. |
| `collect_dep_member_paths` | `collect_dependency_selections` | It records attr-paths *selected* on a dependency argument (`python3Packages.foo`), not "members." Each entry is a selection path. |
| `resolve_dep` | `try_resolve_dependency_argument` | Tries to resolve a formal argument (with any selected members) to a local package file; `try_`/`Option` says it may resolve to nothing. |
| `roots` (throughout) | `root_attributes` | `roots` currently means the catalog attr-path root parameters (`catalogs`); once the graph type exists, "root" reads as "entry package." `root_attributes` names what they are. |

`FileInfo` stays — it's the per-file scan result.

**Track references as** `CatalogRef` **from the scan result onward**

Change `FileInfo.refs` from `BTreeSet<String>` to `BTreeSet<CatalogRef>`, converting once where `FileInfo` is built (`analyze_file_at` return). Everything downstream — the closure, `references()`, `scan_package` — is then `CatalogRef`, and the `.map(CatalogRef)` at `mod.rs:199` disappears.

The walker's internal accumulator stays `String`: it does dotted-path surgery (`append_star` → `catalogs.*`, `join_path`, root rewriting, the `format!("{root}.*")` sentinel) that a newtype adds nothing to. The scan-result boundary is the earliest point where the value stops being manipulated as a string. The only extra conversion is the import merge (`rewrite_root` on a child's rewritten refs), which round-trips through the existing `From<CatalogRef> for String` / `From<String>` impls — cheap, or fold it into `rewrite_root`'s signature.

**Make** `PackageGraph` **the center of the computation**

Wrap the ad-hoc `db: HashMap<String, FileInfo>` in a type that owns its keyextension from reference extraction (both currently in `collect_transitive`):

```rust
let mut graph = PackageGraph::new(base_dir, root_attributes);
graph.add_root(rel_file)?;      // read + analyze an entry package
graph.expand_closure()?;        // grow to the closure of dependency args
let references = graph.references();  // BTreeSet<CatalogRef>
```

* `PackageGraph { base_dir, root_attributes, scans: HashMap<String, FileInfo> }` — doc the key format (dependency attr-path joined with `/`, matching `try_resolve_dependency_argument`) and that `scans` is a lazy cache of reac
* `expand_closure` = today's `collect_transitive` traversal minus the ref accumulator; it only follows `dependency_args`.                                                       - `references` unions `refs` over every node — valid because `expand_closurable packages, so the running `result\` set disappears.
* `scan_package{,_with_roots}` become thin wrappers over the four-line sequence; `add_root` is callable more than once, so a multi-target caller like `bin/lock.rs` could later build one graph instead of one per file.

The three `collect_transitive` unit tests seed/read internals in-module, soring against `refset(...)` now that the output is `CatalogRef`); the `.deps`access in `dep_arguments_collected_through_wrapped_lambda` becomes `.dependency_args`.
